astropy/table/column.py

```diff
@@ -222,21 +218,21 @@ def _convert_sequence_data_to_array(data, dtype=None):
             category=FutureWarning,
             message=".*Promotion of numbers and bools to strings.*",
         )
+
+        has_unit = has_len_gt0 and any(hasattr(v, "unit") for v in data)
+
         try:
-            np_data = np.array(data, dtype=dtype)
+            if has_unit:
+                np_data = Quantity(data, dtype)
+            else:
+                np_data = np.array(data, dtype=dtype)
```

Comment:
This is a good place for the ternary operator. Interestingly, it seems like there was a latent bug in the original code calling `Quantity(data, dtype)` since the signature is `Quantity(value, unit, dtype, ...)`.
```suggestion
            cls = Quantity if has_unit else np.array
            np_data = cls(data, dtype=dtype)
```